File: curriculum/units/M02/code/typescript/eventStream.ts
export type HarnessEvent =
  | { type: 'run.started'; runId: string }
  | { type: 'model.delta'; text: string }
  | { type: 'tool.progress'; toolUseId: string; percent: number }
  | { type: 'run.completed'; runId: string }

export type RunSummary = {
  runId: string
  yielded: number
}

export type StreamOptions = {
  failAfterDelta?: boolean
}

export async function* runEventStream(
  runId: string,
  trace: string[],
  options: StreamOptions = {},
): AsyncGenerator<HarnessEvent, RunSummary, void> {
  let yielded = 0
  trace.push('producer.started')
  try {
    trace.push('producer.before:start')
    yielded += 1
    yield { type: 'run.started', runId }
    trace.push('producer.after:start')

    trace.push('producer.before:delta')
    yielded += 1
    yield { type: 'model.delta', text: 'hello' }
    trace.push('producer.after:delta')

    if (options.failAfterDelta) {
      throw new Error('scripted producer failure')
    }

    trace.push('producer.before:complete')
    yielded += 1
    yield { type: 'run.completed', runId }
    trace.push('producer.after:complete')
    return { runId, yielded }
  } finally {
    trace.push('producer.finally')
  }
}

export async function* delegatedRun(
  runId: string,
  trace: string[],
): AsyncGenerator<HarnessEvent, RunSummary, void> {
  trace.push('delegate.started')
  try {
    const summary = yield* runEventStream(runId, trace)
    trace.push(`delegate.return:${summary.yielded}`)
    return summary
  } finally {
    trace.push('delegate.finally')
  }
}

export class PushAsyncQueue<T> implements AsyncIterableIterator<T> {
  private readonly queue: T[] = []
  private waiting?: {
    resolve: (result: IteratorResult<T>) => void
    reject: (error: unknown) => void
  }
  private closed = false
  private failure: unknown

  get bufferedCount(): number {
    return this.queue.length
  }

  [Symbol.asyncIterator](): AsyncIterableIterator<T> {
    return this
  }

  next(): Promise<IteratorResult<T>> {
    if (this.queue.length > 0) {
      return Promise.resolve({ done: false, value: this.queue.shift()! })
    }
    if (this.failure !== undefined) return Promise.reject(this.failure)
    if (this.closed) return Promise.resolve({ done: true, value: undefined })
    return new Promise((resolve, reject) => {
      this.waiting = { resolve, reject }
    })
  }

  enqueue(value: T): void {
    if (this.closed) throw new Error('queue is closed')
    if (this.waiting) {
      const waiting = this.waiting
      this.waiting = undefined
      waiting.resolve({ done: false, value })
      return
    }
    this.queue.push(value)
  }

  done(): void {
    this.closed = true
    this.waiting?.resolve({ done: true, value: undefined })
    this.waiting = undefined
  }

  error(error: unknown): void {
    this.failure = error
    this.waiting?.reject(error)
    this.waiting = undefined
  }

  return(): Promise<IteratorResult<T>> {
    this.closed = true
    this.queue.length = 0
    return Promise.resolve({ done: true, value: undefined })
  }
}

File: curriculum/units/M02/code/python/event_stream.py
from __future__ import annotations

from dataclasses import dataclass
from typing import AsyncIterator, Literal


@dataclass(frozen=True)
class HarnessEvent:
    type: Literal["run.started", "model.delta", "run.completed"]
    value: str


async def run_event_stream(
    run_id: str, trace: list[str], fail_after_delta: bool = False
) -> AsyncIterator[HarnessEvent]:
    trace.append("producer.started")
    try:
        trace.append("producer.before:start")
        yield HarnessEvent("run.started", run_id)
        trace.append("producer.after:start")
        trace.append("producer.before:delta")
        yield HarnessEvent("model.delta", "hello")
        trace.append("producer.after:delta")
        if fail_after_delta:
            raise RuntimeError("scripted producer failure")
        trace.append("producer.before:complete")
        yield HarnessEvent("run.completed", run_id)
        trace.append("producer.after:complete")
    finally:
        trace.append("producer.finally")


class PushAsyncQueue:
    def __init__(self) -> None:
        self._queue: list[int] = []
        self._closed = False

    @property
    def buffered_count(self) -> int:
        return len(self._queue)

    def enqueue(self, value: int) -> None:
        if self._closed:
            raise RuntimeError("queue is closed")
        self._queue.append(value)

    def done(self) -> None:
        self._closed = True

    def __aiter__(self) -> PushAsyncQueue:
        return self

    async def __anext__(self) -> int:
        if self._queue:
            return self._queue.pop(0)
        if self._closed:
            raise StopAsyncIteration
        raise RuntimeError("teaching queue has no pending waiter implementation")
